### ISSUE TITLE
Handling "Magnitude of gradient is bad" in Python

### ISSUE THREAD

**Issue by danielhers** [clab/dynet#515]
A runtime error is raised [here](https://github.com/clab/dynet/blob/master/dynet/training.cc#L69) if the gradient is `nan` or `inf` in `update()`.

Previous discussions of this issue: [here](https://groups.google.com/forum/#!topic/dynet-users/ety7ZU3DNYc) and [here](https://groups.google.com/forum/#!topic/dynet-users/RLCUI0073dk). In the latter, Matt suggests wrapping the call to `update()` in a `try ... catch`, but this will only work in C++ and not in Python, since it's not a Python exception being thrown.

I know it might be a problem with hyperparameters, but I would still like to be able to handle this in code (for example when doing hyperparameter search) rather than get thrown out of the program entirely.

I tried getting the gradient and checking it myself:
```
...
loss.backward()
if np.linalg.norm(loss.gradient()) not in (np.inf, np.nan):
        trainer.update()
...
```
but this gives me the following error:
`RuntimeError: Requested gradient for node 169282, but backward pass was computed from node 169282`
coming from [here](https://github.com/clab/dynet/blob/0e4550542825dc7e9ceee1158490f531c49f3e0b/dynet/exec.cc#L42).

Is this a bug? Shouldn't it be possible to get the gradient of a node up to which the backward pass was computed?

**Comment by pmichel31415** [comment]
The 1st one can be solved by passing the c++ exception to python.

The 2nd is indeed a bug.

I'll try and fix this today.

**Comment by danielhers** [comment]
Thanks. By passing the C++ exception to Python, you mean a change in the DyNet code, right?

**Comment by pmichel31415** [comment]
Yes, I'll include it in my commit

**Comment by danielhers** [comment]
Cool, thanks!

**Comment by pmichel31415** [comment]
OK this will be fixed by my latest PR, I'll merge once the tests complete

**Comment by pmichel31415** [comment]
Fixed by #518
